### PR TITLE
Add Support for Switchboard's GameOfNodes voter-weight add-in

### PR DESCRIPTION
**Pull request by S1nus**
Switchboard oracle queues are governed democratically by the nodes on the queue. For instance, admitting a new node to a queue is put up to a vote.

This PR adds code to recognize realms using the add-in, and use it as a VotingClient in those realms.

### CODE REVIEW

File: hooks/useVotingPlugins.ts [hunk 1]
```diff
@@ -3,12 +3,31 @@ import useWalletStore from 'stores/useWalletStore'
...
+  getGovernanceAccount,
+  GovernanceAccountType,
+  Governance
+} from '@solana/spl-governance'
 import { getNftMaxVoterWeightRecord } from 'NftVotePlugin/sdk/accounts'
 import { notify } from '@utils/notifications'
+import { AccountLayout, NATIVE_MINT } from '@solana/spl-token'
+import * as anchor from '@project-serum/anchor'
+import * as sbv2 from '@switchboard-xyz/switchboard-v2'
+import sbidl from 'SwitchboardVotePlugin/switchboard_v2.json';
```
Comment (mgild) [inline hooks/useVotingPlugins.ts:23]: camelCase

File: hooks/useRealm.tsx [hunk 2]
```diff
@@ -47,6 +49,11 @@ export default function useRealm() {
   } = useWalletStore((s) => s.selectedRealm)
   const votingPower = useDepositStore((s) => s.state.votingPower)
   const nftVotingPower = useNftPluginStore((s) => s.state.votingPower)
+  const sbVotingPower = useSwitchboardPluginStore((s) => s.state.votingPower)
+  const [realmInfo, setRealmInfo] = useState<RealmInfo | undefined>(undefined)
+  //getQueuesForSelectedRealm();
```
Comment (mgild) [inline hooks/useRealm.tsx:54]: cleanup

File: hooks/useCreateProposal.ts [hunk 3]
```diff
@@ -42,9 +43,20 @@ export default function useCreateProposal() {
       governance!.account.config
     )
 
+    /*const defaultProposalMint =
```
Comment (mgild) [inline hooks/useCreateProposal.ts:46]: cleanup

File: components/TokenBalance/TokenBalanceCardWrapper.tsx [hunk 4]
```diff
@@ -55,6 +65,17 @@ const TokenBalanceCardWrapper = ({
         </>
       )
     }
+    if (
+      isSwitchboardMode &&
+      (!ownTokenRecord ||
+        ownTokenRecord.account.governingTokenDepositAmount.isZero())
+    ) {
+      return (
+        <>
```
Comment (mgild) [inline components/TokenBalance/TokenBalanceCardWrapper.tsx:74]: empty guard? not all that familiar with react, this intentional?

File: components/TokenBalance/SwitchboardPermissionCard.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+//import useNftPluginStore from 'NftVotePlugin/store/nftPluginStore'
+import { useState, useEffect } from 'react'
+//import useVotePluginsClientStore from 'stores/useVotePluginsClientStore'
+import useWalletStore from 'stores/useWalletStore'
+
+const SwitchboardPermissionCard = () => {
+  const { fmtUrlWithCluster } = useQueryContext()
+  const connected = useWalletStore((s) => s.connected)
+  const wallet = useWalletStore((s) => s.current)
+  /*const client = useVotePluginsClientStore(
```
Comment (mgild) [inline components/TokenBalance/SwitchboardPermissionCard.tsx:26]: cleanup

File: components/TokenBalance/SwitchboardPermissionCard.tsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+import { ChevronRightIcon } from '@heroicons/react/outline'
+import useQueryContext from '@hooks/useQueryContext'
+import useRealm from '@hooks/useRealm'
+//import { NftVoterClient } from '@solana/governance-program-library'
+import {
+  getTokenOwnerRecordAddress,
+  //SYSTEM_PROGRAM_ID,
+  //withCreateTokenOwnerRecord,
+} from '@solana/spl-governance'
+//import { Transaction, TransactionInstruction } from '@solana/web3.js'
```
Comment (mgild) [inline components/TokenBalance/SwitchboardPermissionCard.tsx:13]: cleanup

File: actions/createProposal.ts [hunk 7]
```diff
@@ -201,6 +228,10 @@ export const createProposal = async (
     transaction1.add(...prerequisiteInstructions, ...instructions)
     transaction2.add(...insertInstructions)
 
+    console.log("the two transactions");
```
Comment (mgild) [inline actions/createProposal.ts:231]: remove

File: actions/createProposal.ts [hunk 8]
```diff
@@ -108,6 +114,8 @@ export const createProposal = async (
     tokenOwnerRecord,
     'createProposal'
   )
+  console.log("Plugin");
```
Comment (mgild) [inline actions/createProposal.ts:117]: remove

File: actions/createProposal.ts [hunk 9]
```diff
@@ -86,16 +86,22 @@ export const createProposal = async (
 
   // sum up signers
   const signers: Keypair[] = instructionsData.flatMap((x) => x.signers ?? [])
+  console.log("the signers:");
+  console.log(signers);
   const shouldSplitIntoSeparateTxs: boolean = instructionsData
     .flatMap((x) => x.shouldSplitIntoSeparateTxs)
     .some((x) => x)
 
   // Explicitly request the version before making RPC calls to work around race conditions in resolving
```
Comment (mgild) [inline actions/createProposal.ts:95]: cleanup

File: actions/createProposal.ts [hunk 10]
```diff
@@ -86,16 +86,22 @@ export const createProposal = async (
 
   // sum up signers
   const signers: Keypair[] = instructionsData.flatMap((x) => x.signers ?? [])
+  console.log("the signers:");
```
Comment (mgild) [inline actions/createProposal.ts:89]: remove

File: actions/castVote.ts [hunk 11]
```diff
@@ -41,10 +41,15 @@ export async function castVote(
   const payer = walletPubkey
   // Explicitly request the version before making RPC calls to work around race conditions in resolving
   // the version for RealmInfo
-  const programVersion = await getGovernanceProgramVersion(
+
```
Comment (mgild) [inline actions/castVote.ts:44]: remove

File: hooks/useVotingPlugins.ts [hunk 12]
```diff
@@ -92,6 +119,152 @@ export function useVotingPlugins() {
...
+            oracleData.oracleAuthority.equals(wallet.publicKey)
+          ) {
+            myNodesForRealm.push(oracle);
+            let [p] = sbv2.PermissionAccount.fromSeed(
+              switchboardProgram,
+              queueAuthority,
+              queuePk,
+              oracle
+            );
+            //let ix = await p.setVoterWeightTx({govProgram: realm.owner, pubkeySigner: wallet.publicKey}, addinProgram, grantAuthority);
```
Comment (mgild) [inline hooks/useVotingPlugins.ts:204]: remove

File: models/voteWeights.ts [hunk 13]
```diff
@@ -314,6 +314,107 @@ export class VoteNftWeight implements VoterWeightInterface {
   }
 }
 
+export class SwitchboardQueueVoteWeight implements VoterWeightInterface {
+  communityTokenRecord: ProgramAccount<TokenOwnerRecord> | undefined
+  //councilTokenRecord: ProgramAccount<TokenOwnerRecord> | undefined
```
Comment (mgild) [inline models/voteWeights.ts:319]: cleanup

File: models/voteWeights.ts [hunk 14]
```diff
@@ -314,6 +314,107 @@ export class VoteNftWeight implements VoterWeightInterface {
...
+    throw new Error('Current wallet has no Token Owner Records')
+  }
+
+  hasMinCommunityWeight(minCommunityWeight: BN) {
+    return (
+      this.communityTokenRecord && this.votingPower.cmp(minCommunityWeight) >= 0
+    )
+  }
+  hasMinCouncilWeight(minCouncilWeight: BN) {
+    /*return (
```
Comment (mgild) [inline models/voteWeights.ts:354]: cleanup

File: models/voteWeights.ts [hunk 15]
```diff
@@ -314,6 +314,107 @@ export class VoteNftWeight implements VoterWeightInterface {
...
+  }
+  canCreateGovernanceUsingCommunityTokens(realm: ProgramAccount<Realm>) {
+    return true
+    return this.hasMinCommunityWeight(
+      realm.account.config.minCommunityTokensToCreateGovernance
+    )
+  }
+  canCreateGovernanceUsingCouncilTokens() {
+    return false
+    /*return true
```
Comment (mgild) [inline models/voteWeights.ts:374]: cleanup

File: models/voteWeights.ts [hunk 16]
```diff
@@ -314,6 +314,107 @@ export class VoteNftWeight implements VoterWeightInterface {
...
+    if (isCouncil) {
+      return !this.councilTokenRecord?.account.governingTokenDepositAmount.isZero()
+    }
+    if (isCommunity) {
+      return !this.votingPower.isZero()
+    }
+  }
+
+  getTokenRecordToCreateProposal(config: GovernanceConfig) {
+    // Prefer community token owner record as proposal owner
```
Comment (mgild) [inline models/voteWeights.ts:404]: cleanup

File: actions/castVote.ts (hunk 11 above)
Comment (S1nus) [reply actions/castVote.ts:44]: removed- will need to restore when i want to test on localnet.

File: hooks/useCreateProposal.ts [hunk 17]
```diff
@@ -44,7 +45,8 @@ export default function useCreateProposal() {
 
     const defaultProposalMint =
       !mint?.supply.isZero() ||
-      realm?.account.config.useMaxCommunityVoterWeightAddin
+      realm?.account.config.useMaxCommunityVoterWeightAddin ||
+      true
```
Comment (abrzezinski94) [inline hooks/useCreateProposal.ts:49]: why true in here ?

File: stores/useGovernanceAssetsStore.tsx [hunk 18]
```diff
@@ -116,9 +119,11 @@ const useGovernanceAssetsStore = create<GovernanceAssetsStore>((set, _get) => ({
     set((s) => {
       s.loadGovernedAccounts = false
     })
+
     const governancesArray = _get().governancesArray.filter(
       (x) => x.pubkey.toBase58() === governancePk.toBase58()
     )
+    console.log(governancesArray)
```
Comment (abrzezinski94) [inline stores/useGovernanceAssetsStore.tsx:126]: nit: remove console logs

File: utils/uiTypes/VotePlugin.ts [hunk 19]
```diff
@@ -191,6 +203,28 @@ export class VotingClient {
...
       }
     }
+    if (this.client instanceof SwitchboardQueueVoterClient) {
+      /*console.log("It's a switchboard client.")
+      console.log('the realm is')
+      console.log(this.realm)
+      console.log('the wallet is')
+      console.log(this.walletPk)
+      console.log("MY ORACLES");*/
+      console.log("Vote client logging oracles and instructions");
```
Comment (abrzezinski94) [inline utils/uiTypes/VotePlugin.ts:213]: nit: remove console logs

**Review by abrzezinski94 — APPROVED** [review]
Hello please remove unnecessary console logs, use package.json and yarn.lock from master with only changes that you need.

File: hooks/useCreateProposal.ts (hunk 17 above)
Comment (S1nus) [reply hooks/useCreateProposal.ts]: I think `useMaxCommunityVoterWeightAddin` is a mistake, so I have corrected it to `useCommunityVoterWeightAddin`. Thoughts?

File: yarn.lock [hunk 20]
```diff
@@ -2,314 +2,164 @@
 # yarn lockfile v1
```
[the remark this answers is not in the document]
Comment (S1nus) [reply yarn.lock]: just copied from master. hopefully that's ok

File: hooks/useCreateProposal.ts (hunk 17 above)
Comment (abrzezinski94) [reply hooks/useCreateProposal.ts]: yes you are right version with useCommunityvoterWeightAddin will be better in most plugin cases please replace it in this files too.

Zrzut ekranu 2022-06-11 o 21 38 59

**Pull request merged**